grub legacy: remove

ajs124 opened the pull request:
###### Description of changes
It's been a while, maybe now is finally the time to get rid of this?

If you're still using this and _need_ it, please tell me why.

###### Things done
- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

ajs124 committed:
grub1: remove

ajs124 committed:
grub2: move file to default

ajs124 committed:
nixos/installer/grub1: remove

ajs124 committed:
nixos/*: remove boot.grub.version

samueldr commented on nixos/modules/system/boot/loader/grub/grub.nix:
```diff
@@ -843,6 +814,7 @@ in
 
         See the boot.initrd.secrets option documentation for more information.
       '')
+      (mkRemovedOptionModule [ "boot" "loader" "grub" "version" ] "Support for version 0.9x of GRUB was removed after being unsupported upstream for around a decade")
```
The option could also have a type of `lib.types.enum [ 2 ]` while in the transition period.

samueldr replied:
Scratch that, make it `types.enum [ 1 2 ]` and add a `config.assertion` that adds a user-actionable message.

ajs124 replied:
Oh, I didn't think of that… I'll implement one of the suggestions or try to come up with something else to not cause churn, once we figure out if we don't need to keep this to support the AWS overlords decisions.

samueldr replied:
We probably will want to do `types.enum [ 1 2 ]` either way, but *somehow* only keep the config generation codepath alive for grub 1; e.g. still dropping all packages and all other use cases except config generation.

It might even be worth it to split the grub 1 and 2 config generations? I'm not even sure it really needed to be implemented through the same script, they are different enough that we should consider them two distinct bootloaders.

samueldr commented:
Reviewing the changes, I'm reminded about how some virtualization software boots through a host-controlled GRUB. This is why `boot.loader.grub.version = if cfg.hvm then 2 else 1;` was used in the `amazon-image.nix.

To rephrase this, AFAIUI some virtualization software will consume the grub (1) config, but the guest is not in control of the GRUB install.

An "AWS expert" advice may be needed here to better understand the issue.

ajs124 commented:
> An "AWS expert" advice may be needed here to better understand the issue.

Do we have one of those?

exarkun commented:
> > An "AWS expert" advice may be needed here to better understand the issue.
> 
> Do we have one of those?

I am *not* an AWS expert.  Also, I am *not* a grub expert.

I do use NixOS on AWS, though.  I am somewhat familiar with two virtualization schemes AWS offers - PV and HVM.

There used to be more reason to prefer PV over HVM but nowadays I usually simplify the comparison to  "HVM is better".

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/UserProvidedKernels.html#Paravirtual_instances explains some of the details related to how PV instances are booted.  According to that page, PV instances are booted with a "patched GNU GRUB 0.97".

If the question is "Will *any* AWS configuration break if NixOS forces GRUB 2?" then I'm pretty sure the answer is yes.

If NixOS is willing to break some configurations in order to simplify maintenance or realize some other development benefits, some relevant facts to consider are...

* PV used to perform some workloads faster than HVM but if any such cases remain now they are very obscure
* All instance types are available in all data centers as HVM and only some are available in some data centers as PV
* AMIs are created as PV AMIs or HVM AMIs (so if NixOS is not offering PV AMIs then GRUB 1 is already irrelevant as far as the "official" NixOS AMIs are concerned)
* Amazon [recommends HVM](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/virtualization_types.html):
  > For the best performance, we recommend that you use current generation instance types and HVM AMIs when you launch your instances

Hope this helps.

RaitoBezarius commented:
I think it's time to get this merged @ajs124

ajs124 commented:
@GrahamcOfBorg test grub installer.simple installer.simpleUefiGrub

ajs124 committed:
nixos/grub: turn definition of version option from an error into a warning

because a lot of configurations (generated by nixos-generate-config) contain it

ajs124 commented:
`nixosTests.grub` already fails on aarch64-linux on master

RaitoBezarius reviewed (approved):
Oh yes <3

RaitoBezarius commented:
Let's go?

RaitoBezarius commented:
@Mic92 this is party time on Friday

SuperSandro2000 commented:
I just noticed that some descriptions still contained v1 references and I removed them in #238822
